core/services/llo/observation/data_source.go

```diff
@@ -85,139 +100,285 @@ func NewDataSource(lggr logger.Logger, registry Registry, t Telemeter) llo.DataS
+	{
+		// Update the list of streams to observe for this config digest and set the timeout
+		d.configDigestToStreamMu.Lock()
+		// StreamValues  needs a copy to avoid concurrent access
+		d.setObservableStreams(ctx, streamValues, opts)
+		d.configDigestToStreamMu.Unlock()
+
+		if !d.observationLoopStarted.Load() {
+			loopStartedCh := make(chan struct{})
+			go d.startObservationLoop(loopStartedCh)
+			<-loopStartedCh
+		}
+	}
 
-	if opts.VerboseLogging() {
-		streamIDs := make([]streams.StreamID, 0, len(streamValues))
-		for streamID := range streamValues {
-			streamIDs = append(streamIDs, streamID)
+	// Fetch the cached observations for all streams.
+	for streamID := range streamValues {
+		val := d.fromCache(streamID)
+		if val != nil {
+			streamValues[streamID] = val
 		}
-		sort.Slice(streamIDs, func(i, j int) bool { return streamIDs[i] < streamIDs[j] })
-		lggr = logger.With(lggr, "streamIDs", streamIDs)
-		lggr.Debugw("Observing streams")
 	}
 
-	var wg sync.WaitGroup
-	wg.Add(len(streamValues))
+	return nil
+}
 
-	var mu sync.Mutex
-	successfulStreamIDs := make([]streams.StreamID, 0, len(streamValues))
-	var errs []ErrObservationFailed
+func (d *dataSource) setObservableStreams(ctx context.Context, streamValues llo.StreamValues, opts llo.DSOpts) {
+	values := make(llo.StreamValues, len(streamValues))
+	for streamID := range streamValues {
+		values[streamID] = nil
+	}
 
-	// oc only lives for the duration of this Observe call
-	oc := NewObservationContext(lggr, d.registry, d.t)
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(100 * time.Millisecond)
+	}
 
-	// Telemetry
-	{
-		// Size needs to accommodate the max number of telemetry events that could be generated
-		// Standard case might be about 3 bridge requests per spec and one stream<=>spec
-		// Overallocate for safety (to avoid dropping packets)
-		telemCh := d.t.MakeObservationScopedTelemetryCh(opts, 10*len(streamValues))
-		if telemCh != nil {
-			if d.t.CaptureEATelemetry() {
-				ctx = pipeline.WithTelemetryCh(ctx, telemCh)
+	streams := make(llo.StreamValues)
+	for streamID := range values {
+		streams[streamID] = values[streamID]
+	}
+
+	d.configDigestToStream[opts.ConfigDigest()] = observableStreamValues{
+		opts:                opts,
+		streamValues:        streams,
+		observationInterval: time.Until(deadline),
+	}
+}
+
+// startObservationLoop continuously makes observations for the streams in d.configDigestToStream and stores those in
+// the cache. It does not check for cached versions, it always calculates fresh values.
+//
+// NOTE: This method needs to be run in a goroutine.
+func (d *dataSource) startObservationLoop(loopStartedCh chan struct{}) {
+	var elapsed time.Duration
+
+	for {
+		select {
+		case <-d.observationLoopCloseCh:
+			return
+		default:
+		}
+
+		loopStart := time.Now()
+		opts, streamValues, observationInterval := d.getObservableStreams()
+
+		ctx, cancel := context.WithTimeout(context.Background(), observationInterval)
```

Comment:
If you use a `services.StopChan` then you can link the context:
```suggestion
	ctx, cancel := d.stopCh.NewCtx()
	defer cancel()
	for {
		if ctx.Err() != nil {
			return
		}

		loopStart := time.Now()
		opts, streamValues, observationInterval := d.getObservableStreams()

		ctx, cancel := context.WithTimeout(ctx, observationInterval)
```

Reply:
This could have been useful, if we were using the context for control. We are using it only as a timeout mechanism, though.

Reply:
I don't understand what you mean. It is important for all service.Service implementations to shut down quickly when Close() is called. Why not connect these?

Reply:
Sure, connected.